core/src/scopyhomeaddpage.cpp

```diff
@@ -1,68 +1,306 @@
 	QString uri = ui->editUri->text();
 	struct iio_context* ctx = ContextProvider::GetInstance()->open(uri);
-	if(ctx) {		
+	if(ctx) {
+		getContextData(ctx);
 		ContextProvider::GetInstance()->close(uri);
 		return true;
 	}	
 	return false;
-
-}
-
-void ScopyHomeAddPage::add() {
-	QString uri = ui->editUri->text();
-	pendingUri = uri;
-	Q_EMIT requestAddDevice("iio",uri);
-
 }
 
 void ScopyHomeAddPage::futureverify()
 {
+	removePluginsCheckBoxs();
+	deviceInfoPage->clear();
 	QFuture<bool> f = QtConcurrent::run(std::bind(&ScopyHomeAddPage::verify,this));
 	fw->setFuture(f);
 }
 
+void ScopyHomeAddPage::futureScan()
+{
+	QString scanParams = scanParamsList.join("");
+	QtConcurrent::run([=](){
+		scanTask->setScanParams(scanParams);
+		scanTask->run();
+	});
+}
+
 void ScopyHomeAddPage::deviceAddedToUi(QString id)
 {
 	if(!pendingUri.isEmpty()) {
+		ui->stackedWidget->setCurrentWidget(ui->deviceDetectionPage);
+		ui->editUri->clear();
 		Q_EMIT requestDevice(id);
 		pendingUri = "";
 	}
 }
 
-ScopyHomeAddPage::~ScopyHomeAddPage()
+void ScopyHomeAddPage::scanFinished(QStringList scanCtxs)
 {
-	delete ui;
+	ui->comboBoxContexts->clear();
+	ui->uriMessageLabel->clear();
+	for (QString ctx: scanCtxs) {
+		ui->comboBoxContexts->addItem(ctx);
+	}
+	findAvailableSerialPorts();
+	ui->editUri->clear();
+	ui->editUri->setText(ui->comboBoxContexts->currentText());
+}
+
+void ScopyHomeAddPage::deviceLoaderInitialized()
+{
+	QList<Plugin *> plugins = deviceImpl->plugins();
+	for (Plugin *p : qAsConst(plugins)) {
+		QWidget *pluginInfo = new QWidget();
+		QHBoxLayout *hLayout = new QHBoxLayout(pluginInfo);
+
+		QCheckBox *cb = new QCheckBox();
+		cb->setText(p->name());
+		cb->setChecked(p->enabled());
+		connect(cb, &QCheckBox::toggled, this, [=](bool en){
+			p->setEnabled(en);
+		});
+
+		QLabel *descriptionLabel = new QLabel();
+		descriptionLabel->setWordWrap(true);
+		descriptionLabel->setText(p->description());
+
+		pluginInfo->setStyleSheet("background-color: rgb(0, 0, 0);");
+		hLayout->setContentsMargins(5, 5, 5, 5);
+		hLayout->addWidget(cb);
+		hLayout->addWidget(descriptionLabel);
+		hLayout->setStretch(0,1);
+		hLayout->setStretch(1,3);
+		ui->subSeparatorPlugins->getContentWidget()->layout()->addWidget(pluginInfo);
+	}
+	ui->stackedWidget->setCurrentWidget(ui->deviceInfoPage);
+}
+
+void ScopyHomeAddPage::createDevice()
+{
+	QString uri = ui->editUri->text();
+	deviceImpl = DeviceFactory::build(uri, pluginManager, "");
+	DeviceLoader* dl = new DeviceLoader(deviceImpl, this);
+	dl->init();
+
+	connect(dl, &DeviceLoader::initialized, this, &ScopyHomeAddPage::deviceLoaderInitialized);
+	connect(dl, &DeviceLoader::initialized, dl, &QObject::deleteLater); // don't forget to delete loader once we're done
+}
+
+void ScopyHomeAddPage::removePluginsCheckBoxs()
+{
+	if (ui->subSeparatorPlugins->getContentWidget()->layout()) {
+		QLayoutItem* item;
+		while ((item = ui->subSeparatorPlugins->getContentWidget()->layout()->takeAt(0)) != nullptr) {
```

Comment:
will this throw index out of range at some point?

Reply:
It should do this when you delete all the items from a layout.